Products/CMFCore/indexing.py

```diff
@@ -0,0 +1,330 @@
+# -*- coding: utf-8 -*-
+from Acquisition import aq_base, aq_inner, aq_parent
+from logging import getLogger
+from Products.CMFCore.interfaces import IIndexQueue
+from Products.CMFCore.interfaces import IIndexQueueProcessor
+from Products.CMFCore.interfaces import InvalidQueueOperation
+from Products.CMFCore.interfaces import IPortalCatalogQueueProcessor
+from Products.CMFCore.utils import getToolByName
+from threading import local
+from transaction import get as getTransaction
+from transaction.interfaces import ISavepointDataManager
+from zope.component import getSiteManager
+from zope.interface import implementer
+from zope.publisher.interfaces.browser import IBrowserRequest
+
+
+# constants for indexing operations
+UNINDEX = -1
+REINDEX = 0
+INDEX = 1
+
+logger = getLogger('Products.CMFCore.indexing')
+debug = logger.debug
+
+localQueue = None
+processing = set()
+
+
+@implementer(IPortalCatalogQueueProcessor)
+class PortalCatalogProcessor(object):
+
+    def index(self, obj, attributes=None):
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        catalog._indexObject(obj)
+
+    def reindex(self, obj, attributes=None, update_metadata=1):
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        catalog._reindexObject(obj, idxs=attributes, update_metadata=update_metadata)
+
+    def unindex(self, obj):
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        catalog._unindexObject(obj)
+
+    def begin(self):
+        pass
+
+    def commit(self):
+        pass
+
+    def abort(self):
+        pass
+
+    @staticmethod
+    def get_dispatcher(obj, name):
+        """ return named indexing method according on the used mixin class """
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        if catalog is None:
```

Comment:
Really? I myself definitely prefer an early return.
For me it is much clearer: the method exits here, instead of possibly executing one final command 20 lines down, for which you need to carefully inspect the indentation.
And it avoids multiple indents that would result in a much shorter available line length, which in turn would result in more wrapped lines.
But for methods with four lines it does not matter to me.